Introduce TemplateRenderer for prompt templating

ThomasVitale opened the pull request:
- Introduce new TemplateRenderer API providing the logic for rendering an input template.
- Update the PromptTemplate API to accept a TemplateRenderer object at construction time.
- Move ST logic to StTemplateRenderer implementation, used by default in PromptTemplate. Additionally, make start and end delimiter character configurable.
- Extend ChatClient API to support passing a custom TemplateRenderer.

Relates to gh-2655

![Untitled-2025-04-16-0204](https://github.com/user-attachments/assets/eb52fe28-2d1a-4f16-9462-ca5efcdf9275)

markpollack commented:
This looks very nice, thank you very much.  I need to try it out in practice in particular what is the current impact when using out of the box advisors like QuestionAnswer advisor.  

The other half of the equation in this area is around resource loading of the appropriate template text based on the template renderer and how that may percolate through the APIs.  For example well known filename/resource-name extensions get looked up automatically based on a prefix/suffix combo 

`question-answer-advisor.st` is the default (though I know you would like to keep that as a final static string - maybe there is a new resource type for a static field we can use) and `question-answer-advisor.mustache`

ThomasVitale commented:
Thanks for the review!

About Advisors and other components that internally rely on prompts (e.g. some of the modular RAG components), the strategy I'd recommend is for them to accept a PromptTemplate instead of a String to customize the prompt/rendering. 

Actually, the modular RAG components are already doing that. So, with the changes in this PR, it's already possible for them to use a different templating strategy because they all accept a PromptTemplate instance. I would recommend QuestionAnswerAdvisor to do the same.

The other issue I think can be handled independently from the templating: loading prompts from external sources (classpath, Spring Cloud Config Server, HTTP endpoint, data store...). I would keep that logic outside PromptTemplate, reducing its responsibility to "just" take a template and render it. A separate API could provide the functionality to load prompts templates from different sources.

ThomasVitale commented:
I can share some thoughts about that use case later. I've been thinking about it a bit.

ThomasVitale commented:
I added support for PromptTemplate in QuestionAnswerAdvisor and added integration tests to showcase how it would work: spring-ai-integration-tests/src/test/java/org/springframework/ai/integration/tests/client/advisor/QuestionAnswerAdvisorIT.java

Here's how to customize prompts in RetrievalAugmentationAdvisor (already supported): https://github.com/ThomasVitale/spring-ai/pull/1/files#diff-ed44486aec8a4d37f17bf82984d4f06342da34729c506cb61336cfb13ce3fd50R266

In both examples, I'm customising the prompt content and also the rendering strategy (using "<variable>" instead of "{variable}").
